# Introdução ao Pandas
# Séries

#### 1 - Importar o Pandas

In [1]:
import pandas as pd
pd.Series?

#### 2 - Lista de strings em Série

In [2]:
animals = ['Gato', 'Cachorro', 'Rato']
pd.Series(animals)

0        Gato
1    Cachorro
2        Rato
dtype: object

#### 3 - Lista de inteiros em Série

In [3]:
numbers = [1,2, None]
pd.Series(numbers)

0    1.0
1    2.0
2    NaN
dtype: float64

#### 4 - None vs. NAN

In [4]:
import numpy as np

animals = ['Tiger', 'Bear', None]
pd.Series(animals)

0    Tiger
1     Bear
2     None
dtype: object

In [5]:
numbers = [1, 2, None]
pd.Series(numbers)

0    1.0
1    2.0
2    NaN
dtype: float64

#### 5 - NAN não é None

In [6]:
import numpy as np
np.nan == None

False

#### 6 - NAN == NAN é falso

In [7]:
np.nan == np.nan

False

#### 7 - É NAN?

In [8]:
np.isnan(np.nan)

True

#### 8 - Séries a partir de dicionários

In [9]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

#### 9 - Objeto do índice

In [10]:
s.index

Index(['Archery', 'Golf', 'Sumo', 'Taekwondo'], dtype='object')

#### 10 - Índice como uma lista explícita para a série

In [11]:
s = pd.Series(['Gato', 'Cachorro', 'Rato'], index = ['Vitor', 'Pedro', 'Santiago'])
s


Vitor           Gato
Pedro       Cachorro
Santiago        Rato
dtype: object

#### 11 - Índice como uma lista explícita para a série criada a partir de um dicionário

In [12]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports, index=['Golf', 'Sumo', 'Hockey'])
s

Golf      Scotland
Sumo         Japan
Hockey         NaN
dtype: object

# Consultando Séries

#### 12 - Lista dos esportes em nosso índice e uma lista de países como valores. Criamos um dicionário e assim uma série.


In [13]:
sports = {'Archery': 'Bhutan',
          'Golf': 'Scotland',
          'Sumo': 'Japan',
          'Taekwondo': 'South Korea'}
s = pd.Series(sports)
s

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

#### 13 - Para ver o quarto país usamos o atributo iloc com o parâmetro 3. Para ver qual país tem o golfe como seu esporte nacional, utilizamos o atributo loc com um parâmetro golfe.

In [14]:
s.iloc[3]

'South Korea'

In [15]:
s.loc['Golf']

'Scotland'

#### 14 - Possibilidade de usar o operador de indexação diretamente na série em si.

In [16]:
s[3]

'South Korea'

In [17]:
s['Golf']

'Scotland'

#### 15 - A opção mais segura é ser explícito e usar os atributos iloc ou loc diretamente. Por exemplo....

In [18]:
sports = {99: 'Bhutan',
          100: 'Scotland',
          101: 'Japan',
          102: 'South Korea'}
s = pd.Series(sports)

In [26]:
# s[0]

#### 16 - Rotina que itera sobre todos os itens da série, somando-os para obter um total.


In [27]:
s = pd.Series([100.0, 120.0, 101.0, 3.00])
s

0    100.0
1    120.0
2    101.0
3      3.0
dtype: float64

In [28]:
total = 0
for item in s:
    total += item
print(total)

324.0


#### 17 - Vetorização Numpy

In [29]:
total = np.sum(s)
print (total)

324.0


#### 18 - Criar uma grande série de números aleatórios. 

In [30]:
s = pd.Series(np.random.randint(0,1000,10000))
s.head()

0    106
1      7
2    780
3    646
4    743
dtype: int64

In [31]:
len(s)

10000

#### 19 - "timeit" com nosso código iterativo sem o Numpy.

In [32]:
%%timeit -n 100
total = 0
for item in s:
    total += item

1.63 ms ± 93.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 20 - Agora vamos com a vetorização do Numpy.

In [33]:
%%timeit -n 100
total = np.sum(s)

117 µs ± 48 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 21 - Broadcasting e head()

In [34]:
s[0:1]+=2
s.head()

0    108
1      7
2    780
3    646
4    743
dtype: int64

#### 22 - O Pandas permite a iteração por meio de uma série assim como num dicionário.

In [126]:
for label, value in s.iteritems():
    s.set_value(label, value+2)
s.head()

AttributeError: 'Series' object has no attribute 'set_value'

#### 23 - Vamos cronometrar as duas abordagens.

In [75]:
%%timeit -n 1
s = pd.Series(np.random.randint(0,1000,10000))
for label, value in s.iteritems():
    s.loc[label]= value+2

550 ms ± 63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [76]:
%%timeit -n 1
s = pd.Series(np.random.randint(0,1000,10000))
s+=2

The slowest run took 4.09 times longer than the fastest. This could mean that an intermediate result is being cached.
512 µs ± 303 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


#### 24 - Tipos misturados para os valores de dados ou índice de rótulos não são problema para o Pandas.


In [77]:
s = pd.Series([1,2,3])
s.loc['Animal'] = "Gato"
s

0            1
1            2
2            3
Animal    Gato
dtype: object

#### 25 . Valores dos índices não são sempre exclusivos.

In [78]:
original_sports = pd.Series({'Archery': 'Bhutan',
                             'Golf': 'Scotland',
                             'Sumo': 'Japan',
                             'Taekwondo': 'South Korea'})
cricket_loving_countries = pd.Series(['Australia',
                                      'Barbados',
                                      'Pakistan',
                                      'England'], 
                                   index=['Cricket',
                                          'Cricket',
                                          'Cricket',
                                          'Cricket'])
all_countries = original_sports.append(cricket_loving_countries)

#### 26 - O método append na verdade não muda a série subjacente. Em vez disso, ele retorna uma nova série que é composta das duas juntas. 

In [79]:
original_sports

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

In [80]:
cricket_loving_countries

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

In [81]:
all_countries

Archery           Bhutan
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
Cricket        Australia
Cricket         Barbados
Cricket         Pakistan
Cricket          England
dtype: object

In [82]:
all_countries.loc['Cricket']

Cricket    Australia
Cricket     Barbados
Cricket     Pakistan
Cricket      England
dtype: object

# DataFrame

#### 27 - Vamos colocar esas séries no DataFrame como sendo o primeiro argumento e atribuir aos valores de índices quais foram as lojas onde as compras foram feitas.

In [83]:
import pandas as pd
purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})
df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])

df.head()

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

#### 28 - Se quisermos selecionar os dados associados à loja 'Store 2', apenas consultamos o atributo "loc" com um único parâmetro.

In [84]:
df.loc['Store 2']

Name                  Vinod
Item Purchased    Bird Seed
Cost                    5.0
Name: Store 2, dtype: object

In [85]:
type(df.loc['Store 2'])

pandas.core.series.Series

#### 29 - Se consultarmos os registros da loja 'Store 1', veremos que Chris e Kevin, ambos compraram na mesma loja de suprimentos.

In [86]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5

#### 30 - Se quisermos apenas listar os custos da 'Store 1', bastaria fornecer dois parâmetros para df.loc, um deles o índice da linha e o outro, o nome da coluna.

In [87]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

#### 31 - Podemos transpôr o DataFrame, usando o atributo T maiúsculo, ele troca as linhas por colunas e vice-versa. Depois podemos usar o método '.loc' Não é muito elegante!

In [88]:
df.T

Store 1       Store 1    Store 2
Name               Chris         Kevyn      Vinod
Item Purchased  Dog Food  Kitty Litter  Bird Seed
Cost                22.5           2.5        5.0

In [89]:
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: object

#### 32 - Operador de indexação diretamente no DataFrame para seleção de coluna.

In [90]:
df['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

#### 33 - Encadeamento (Cuidado)

In [91]:
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

#### 34 - Fatiamento no Dataframe.

In [92]:
df.loc[:,['Name', 'Cost']]

Name  Cost
Store 1  Chris  22.5
Store 1  Kevyn   2.5
Store 2  Vinod   5.0

#### 35 - Deletar dados de séries e de DataFrames. A função drop não altera o DataFrame por padrão. Ao invés disso, lhe retorna uma cópia do DataFrame com as linhas informadas removidas. Podemos ver que o nosso DataFrame original ainda está intacto.


In [93]:
df.drop('Store 1')

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

In [94]:
df

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

#### 36 - Fazendo cópias do dataframe.

In [95]:
copy_df = df.copy()
copy_df = copy_df.drop('Store 1')
copy_df

Name Item Purchased  Cost
Store 2  Vinod      Bird Seed   5.0

#### 37 - O drop tem dois interessantes parâmetros opcionais.

In [96]:
df.drop?

#### 38 - Outra forma de deletar.

In [97]:
del df['Name']
df

Item Purchased  Cost
Store 1       Dog Food  22.5
Store 1   Kitty Litter   2.5
Store 2      Bird Seed   5.0

#### 39 - Adicionar uma nova coluna ao DataFrame. 

In [98]:
df['Location'] = None
df

Item Purchased  Cost Location
Store 1       Dog Food  22.5     None
Store 1   Kitty Litter   2.5     None
Store 2      Bird Seed   5.0     None

# Dataframes - Indexando & Carregando

#### 40 - Podemos criar uma série baseada apenas na categoria preço, usando colchetes.

In [99]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

#### 41 - Broadcasting

In [100]:
costs+=2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

In [101]:
df

Item Purchased  Cost Location
Store 1       Dog Food  24.5     None
Store 1   Kitty Litter   4.5     None
Store 2      Bird Seed   7.0     None

#### 42 - Carregando arquivo CSV - dados das olimpíadas.

In [102]:
df = pd.read_csv('./Data/olympics.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

#### 43 - Ignorar a primeira linha, que era composta por nomes das colunas numéricas.

In [103]:
df = pd.read_csv('./Data/olympics.csv', index_col = 0, skiprows=1)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

#### 44 - Pandas armazena uma lista de todas as colunas no atributo ".columns". 

In [104]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

#### 45 - "inplace" para true, então o Pandas atualiza este Dataframe diretamente.

In [105]:
for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold' + col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver' + col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze' + col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#' + col[1:]}, inplace=True) 

df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

# Consultando DataFrames

#### 46 - Mascaramento Booleano.

In [106]:
df['Gold'] > 0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
                                                ...  
Independent Olympic Participants (IOP) [IOP]    False
Zambia (ZAM) [ZAM]                              False
Zimbabwe (ZIM) [ZIM]                             True
Mixed team (ZZX) [ZZX]                           True
Totals                                           True
Name: Gold, Length: 147, dtype: bool

#### 47 - Sobrepor esta máscara no data frame.

In [107]:
only_gold = df.where(df['Gold'] > 0)
only_gold

# Summer    Gold  Silver  \
Afghanistan (AFG)                                  NaN     NaN     NaN   
Algeria (ALG)                                     12.0     5.0     2.0   
Argentina (ARG)                                   23.0    18.0    24.0   
Armenia (ARM)                                      5.0     1.0     2.0   
Australasia (ANZ) [ANZ]                            2.0     3.0     4.0   
...                                                ...     ...     ...   
Independent Olympic Participants (IOP) [IOP]       NaN     NaN     NaN   
Zambia (ZAM) [ZAM]                                 NaN     NaN     NaN   
Zimbabwe (ZIM) [ZIM]                              12.0     3.0     4.0   
Mixed team (ZZX) [ZZX]                             3.0     8.0     5.0   
Totals                                            27.0  4809.0  4775.0   

                                              Bronze    Total  # Winter  \
Afghanistan (AFG)                                NaN      NaN       NaN   
Algeria (ALG)                                    8.0     15.0       3.0   
Argentina (ARG)                                 28.0     70.0      18.0   
Armenia (ARM)                                    9.0     12.0       6.0   
Australasia (ANZ) [ANZ]                          5.0     12.0       0.0   
...                                              ...      ...       ...   
Independent Olympic Participants (IOP) [IOP]     NaN      NaN       NaN   
Zambia (ZAM) [ZAM]                               NaN      NaN       NaN   
Zimbabwe (ZIM) [ZIM]                             1.0      8.0       1.0   
Mixed team (ZZX) [ZZX]                           4.0     17.0       0.0   
Totals                                        5130.0  14714.0      22.0   

                                              Gold.1  Silver.1  Bronze.1  \
Afghanistan (AFG)                                NaN       NaN       NaN   
Algeria (ALG)                                    0.0       0.0       0.0   
Argentina (ARG)                                  0.0       0.0       0.0   
Armenia (ARM)                                    0.0       0.0       0.0   
Australasia (ANZ) [ANZ]                          0.0       0.0       0.0   
...                                              ...       ...       ...   
Independent Olympic Participants (IOP) [IOP]     NaN       NaN       NaN   
Zambia (ZAM) [ZAM]                               NaN       NaN       NaN   
Zimbabwe (ZIM) [ZIM]                             0.0       0.0       0.0   
Mixed team (ZZX) [ZZX]                           0.0       0.0       0.0   
Totals                                         959.0     958.0     948.0   

                                              Total.1  # Games  Gold.2  \
Afghanistan (AFG)                                 NaN      NaN     NaN   
Algeria (ALG)                                     0.0     15.0     5.0   
Argentina (ARG)                                   0.0     41.0    18.0   
Armenia (ARM)                                     0.0     11.0     1.0   
Australasia (ANZ) [ANZ]                           0.0      2.0     3.0   
...                                               ...      ...     ...   
Independent Olympic Participants (IOP) [IOP]      NaN      NaN     NaN   
Zambia (ZAM) [ZAM]                                NaN      NaN     NaN   
Zimbabwe (ZIM) [ZIM]                              0.0     13.0     3.0   
Mixed team (ZZX) [ZZX]                            0.0      3.0     8.0   
Totals                                         2865.0     49.0  5768.0   

                                              Silver.2  Bronze.2  \
Afghanistan (AFG)                                  NaN       NaN   
Algeria (ALG)                                      2.0       8.0   
Argentina (ARG)                                   24.0      28.0   
Armenia (ARM)                                      2.0       9.0   
Australasia (ANZ) [ANZ]                            4.0       5.0   
...                                           

#### 48 - Existem 100 países que tiveram medalhas de ouro nos jogos de verão, há 147 países no total.

In [108]:
only_gold['Gold'].count()

100

In [109]:
df['Gold'].count()

147

#### 49 - Muitas vezes queremos jogar fora as linhas que não têm nenhum dado. Para isso, usar a função dropna.

In [110]:
only_gold = only_gold.dropna()
only_gold

# Summer    Gold  Silver  Bronze    Total  \
Algeria (ALG)                  12.0     5.0     2.0     8.0     15.0   
Argentina (ARG)                23.0    18.0    24.0    28.0     70.0   
Armenia (ARM)                   5.0     1.0     2.0     9.0     12.0   
Australasia (ANZ) [ANZ]         2.0     3.0     4.0     5.0     12.0   
Australia (AUS) [AUS] [Z]      25.0   139.0   152.0   177.0    468.0   
...                             ...     ...     ...     ...      ...   
Venezuela (VEN)                17.0     2.0     2.0     8.0     12.0   
Yugoslavia (YUG) [YUG]         16.0    26.0    29.0    28.0     83.0   
Zimbabwe (ZIM) [ZIM]           12.0     3.0     4.0     1.0      8.0   
Mixed team (ZZX) [ZZX]          3.0     8.0     5.0     4.0     17.0   
Totals                         27.0  4809.0  4775.0  5130.0  14714.0   

                           # Winter  Gold.1  Silver.1  Bronze.1  Total.1  \
Algeria (ALG)                   3.0     0.0       0.0       0.0      0.0   
Argentina (ARG)                18.0     0.0       0.0       0.0      0.0   
Armenia (ARM)                   6.0     0.0       0.0       0.0      0.0   
Australasia (ANZ) [ANZ]         0.0     0.0       0.0       0.0      0.0   
Australia (AUS) [AUS] [Z]      18.0     5.0       3.0       4.0     12.0   
...                             ...     ...       ...       ...      ...   
Venezuela (VEN)                 4.0     0.0       0.0       0.0      0.0   
Yugoslavia (YUG) [YUG]         14.0     0.0       3.0       1.0      4.0   
Zimbabwe (ZIM) [ZIM]            1.0     0.0       0.0       0.0      0.0   
Mixed team (ZZX) [ZZX]          0.0     0.0       0.0       0.0      0.0   
Totals                         22.0   959.0     958.0     948.0   2865.0   

                           # Games  Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 15.0     5.0       2.0       8.0            15.0  
Argentina (ARG)               41.0    18.0      24.0      28.0            70.0  
Armenia (ARM)                 11.0     1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]        2.0     3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]     43.0   144.0     155.0     181.0           480.0  
...                            ...     ...       ...       ...             ...  
Venezuela (VEN)               21.0     2.0       2.0       8.0            12.0  
Yugoslavia (YUG) [YUG]        30.0    26.0      32.0      29.0            87.0  
Zimbabwe (ZIM) [ZIM]          13.0     3.0       4.0       1.0             8.0  
Mixed team (ZZX) [ZZX]         3.0     8.0       5.0       4.0            17.0  
Totals                        49.0  5768.0    5733.0    6078.0         17579.0  

[100 rows x 15 columns]

#### 50 - Não precisamos, de fato, usar a função where explicitamente.

In [111]:
only_gold = df[df['Gold'] > 0]
only_gold

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   
...                             ...   ...     ...     ...    ...       ...   
Venezuela (VEN)                  17     2       2       8     12         4   
Yugoslavia (YUG) [YUG]           16    26      29      28     83        14   
Zimbabwe (ZIM) [ZIM]             12     3       4       1      8         1   
Mixed team (ZZX) [ZZX]            3     8       5       4     17         0   
Totals                           27  4809    4775    5130  14714        22   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   
...                           ...       ...       ...      ...      ...   
Venezuela (VEN)                 0         0         0        0       21   
Yugoslavia (YUG) [YUG]          0         3         1        4       30   
Zimbabwe (ZIM) [ZIM]            0         0         0        0       13   
Mixed team (ZZX) [ZZX]          0         0         0        0        3   
Totals                        959       958       948     2865       49   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480  
...                           ...       ...       ...             ...  
Venezuela (VEN)                 2         2         8              12  
Yugoslavia (YUG) [YUG]         26        32        29              87  
Zimbabwe (ZIM) [ZIM]            3         4         1               8  
Mixed team (ZZX) [ZZX]          8         5         4              17  
Totals                       5768      5733      6078           17579  

[100 rows x 15 columns]

#### 51 - 101 países ganharam uma medalha de ouro em algum momento.

In [112]:
len(df[(df['Gold'] > 0) | (df['Gold.1']>0)])

101

#### 52 - Há algum país que mais de uma medalha de ouro nos jogos Olímpicos de inverno e nenhuma nos jogos Olímpicos de verão.

In [113]:
df[(df['Gold.1'] > 0) & (df['Gold'] == 0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

# Indexando Dataframes

#### 53 - Digamos que não queremos indexar o DataFrame por países, mas em vez disso, indexar pelo número de medalhas de ouro que foram conquistadas nos jogos de verão.

In [114]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

#### 54 - Podemos nos livrar completamente do índice chamando a função reset_index. Ela torna o índice uma coluna e cria um índice default numerado.

In [115]:
df = df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

#### 55 - Carregar o arquivo CSV com dados Censo dos EUA.

In [116]:
df = pd.read_csv('./Data/census.csv')
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  

[5 rows x 100 columns]

#### 56 - Ver uma lista de todos os valores exclusivos de uma determinada coluna.

In [117]:
df['SUMLEV'].unique()

array([40, 50])

#### 57 - Mantendo dados sobre os  condados.

In [118]:
df=df[df['SUMLEV'] == 50]
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   
5          57322              57322            57373  ...          1.807375   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   
5         -1.177622         -1.748766         -2.062535         -1.369970   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  
5     1.859511    -0.848580    -1.402476    -1.577232    -0.884411  

[5 rows x 100 columns]

#### 58 - Reduzir nossa análise de dados para apenas a estimativa de população total e para o total de nascimentos.

In [119]:
columns_to_keep = ['STNAME',
                   'CTYNAME',
                   'BIRTHS2010',
                   'BIRTHS2011',
                   'BIRTHS2012',
                   'BIRTHS2013',
                   'BIRTHS2014',
                   'BIRTHS2015',
                   'POPESTIMATE2010',
                   'POPESTIMATE2011',
                   'POPESTIMATE2012',
                   'POPESTIMATE2013',
                   'POPESTIMATE2014',
                   'POPESTIMATE2015']
df = df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

#### 59 - Note que temos aqui um índice dual, primeiro o nome do estado e em seguida o nome do condado.

In [120]:
df = df.set_index(['STNAME', 'CTYNAME'])
df.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

#### 60 - Se quisermos ver os resultados da população do Condado de Washtenaw.

In [121]:
df.loc['Alabama', 'Autauga County']

BIRTHS2010           151
BIRTHS2011           636
BIRTHS2012           615
BIRTHS2013           574
BIRTHS2014           623
BIRTHS2015           600
POPESTIMATE2010    54660
POPESTIMATE2011    55253
POPESTIMATE2012    55175
POPESTIMATE2013    55038
POPESTIMATE2014    55290
POPESTIMATE2015    55347
Name: (Alabama, Autauga County), dtype: int64

#### 61 - Comparar dois condados.

In [122]:
df.loc[ [('Alabama', 'Autauga County'),
         ('Alabama', 'Baldwin County')] ]

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709